plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/nessie/IcebergNessieCatalogModule.java

```diff
@@ -58,9 +63,46 @@ public static NessieIcebergClient createNessieIcebergClient(IcebergNessieCatalog
         icebergNessieCatalogConfig.getBearerToken()
                 .ifPresent(token -> builder.withAuthentication(BearerAuthenticationProvider.create(token)));
 
-        return new NessieIcebergClient(builder.build(NessieApiV1.class),
+        // default version is inferred by uri.
+        IcebergNessieCatalogConfig.ClientApiVersion apiVersion = icebergNessieCatalogConfig.getClientAPIVersion()
+                .orElseGet(() -> inferVersionFromURI(icebergNessieCatalogConfig.getServerUri().toString()));
+
+        NessieApiV1 api;
+        switch (apiVersion) {
+            case V1:
+                api = builder.build(NessieApiV1.class);
+                break;
+            case V2:
+                api = builder.build(NessieApiV2.class);
+                break;
+            default:
+                throw new IllegalArgumentException("Unknown API version '" + apiVersion + "'");
+        }
+
+        return new NessieIcebergClient(api,
                 icebergNessieCatalogConfig.getDefaultReferenceName(),
                 null,
                 ImmutableMap.of());
     }
+
+    private static IcebergNessieCatalogConfig.ClientApiVersion inferVersionFromURI(String uri)
+    {
+        checkArgument(uri != null, "URI is not specified in the catalog properties");
+
+        // match for uri ending with /v1, /v2 etc
+        Pattern pattern = Pattern.compile("/v(\\d+)$");
```

Comment:
extract pattern to static final field:  `private static final Pattern VERSION_PATTERN = Pattern.compile("/v(\\d+)$");`

then simplify this method to: 
```
    private IcebergNessieCatalogConfig.ClientApiVersion inferVersionFromURI()
    {
        checkArgument(serverUri != null, "URI is not specified in the catalog properties");
        Matcher matcher = VERSION_PATTERN.matcher(serverUri.toString());
        if (!matcher.find()) {
            throw new IllegalArgumentException("URI doesn't end with the version: %s. Please configure `client-api-version` in the catalog properties explicitly.".formatted(uri));
        }

        return switch (matcher.group(1)) {
            case "1" -> ClientApiVersion.V1;
            case "2" -> ClientApiVersion.V2;
            default -> throw new IllegalArgumentException("Unknown API version in the URI: " + matcher.group(1));
        };
    }
```

If this method is defined on the `IcebergNessieCatalogConfig` it can access serverUri directly.

Negating matcher.find() allows to avoid deep nesting which makes code easier to read.